### ISSUE TITLE
default hostname now is `fedora`, used to be `localhost`

### ISSUE THREAD

**Issue by vrutkovs** [coreos/fedora-coreos-tracker#649]
Booted 33.20201006.1.0 from `next` stream on GCP and default hostname is `fedora`:
```
[    2.041577] systemd[1]: No hostname configured.
[    2.042356] systemd[1]: Set hostname to <fedora>.
```

OKD-on-FCOS expects default hostname to be `localhost` (as on AWS).

See https://github.com/openshift/machine-config-operator/pull/2160#issuecomment-710059362

**Comment by bdurrow** [comment]
perhaps related? https://github.com/peckato1/fedora-systemd-package/commit/6eb8bcde288dda39b163e87ee0926f6f30fcad73

**Comment by bdurrow** [comment]
This is what I see in my boot logs:
```
Welcome to [0;38;2;60;110;180mFedora CoreOS 33.20201006.10.2 dracut-050-63.git20200529.fc33 (Initramfs)[0m!

[    2.456683] systemd[1]: No hostname configured.
[    2.457970] systemd[1]: Set hostname to <fedora>.
```

According to the systemd code this indicates that systemd was compiled with `FALLBACK_HOSTNAME` set to fedora perhaps with `-Dfallback-hostname=fedora`
https://github.com/systemd-rhel/rhel-8/blob/45d093a37b6f8c2ceae9bfd090c5265f35413b46/src/core/hostname-setup.c#L42

**Comment by vrutkovs** [comment]
Ways to fix it:
* Update systemd-resolved to use `localhost` - might take time to propagate in the `next` stream
* Update afterburn to replace `fedora` hostname with `localhost`

Seems afterburn fix is the easiest to implement, but not sure if its a good idea to clean up after systemd-resolved

**Comment by jlebon** [comment]
Maybe let's keep discussing in https://github.com/openshift/machine-config-operator/pull/2160 how to proceed before doing anything here?

**Comment by lucab** [comment]
I was a bit puzzled by this bug report, so I checked the behavior of a plain FCOS 33.x on GCP and I can see the hostname properly set from DHCP in initramfs:

```
NetworkManager[490]: <info>  [1603701253.4075] policy: set 'Wired Connection' (ens4) as default for IPv4 routing and DNS
NetworkManager[490]: <info>  [1603701253.4075] policy: set-hostname: set hostname to 'lbruno-temp.c.example.internal' (from DHCPv4)
```

This seems right to me, and it also matches the behavior I'd expect on other platforms.
How does OKD end up having `fedora` instead? Can we see the full initramfs logs to check what's going on?

**Comment by bdurrow** [comment]
Right, okd disables DHCP on GCP:
https://github.com/openshift/machine-config-operator/blob/f2489433faacc32a7ead1caafcf076b943ffc9d8/templates/common/gcp/files/etc-networkmanager-conf.d-hostname.yaml#L12

If it weren't for that I wouldn't personally have a problem with my deployments but I have modified my installer to allow me to set my clusterdomain to the basedomain instead of having to have a 4 segment domain name.

**Comment by lucab** [comment]
I've opened https://bugzilla.redhat.com/show_bug.cgi?id=1892235 for Fedora package owners to explore whether the branding sugar can be applied in a different place or in some other way which would be easier to control at provisioning time.

**Comment by dustymabe** [comment]
We discussed this in [the community meeting today](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/SFQ62KBJMHT2E7W237EPLHEMGTMKQZKR/).

```
13:20:14  dustymabe | #info we'll try to work with systemd/NM to see if "better fixes" for this
                    | issue can be placed upstream in those projects. If agreeable we'll push
                    | for a short term hack in MCO, and try to get builds into Fedora soonish.
                    | If not agreeable then we'll probably go with supporting this in afterburn.

```

@lucab has started the discussion with systemd already in [BZ1892235](https://bugzilla.redhat.com/show_bug.cgi?id=1892235). The discussion with NM is upcoming.

**Comment by darkmuggle** [comment]
https://github.com/coreos/afterburn/pull/512 addresses the setting of the default hostname in GCP

**Comment by darkmuggle** [comment]
Comment https://github.com/coreos/fedora-coreos-tracker/issues/649#issuecomment-718148008 does not address the immediate pain in GCP. 

On GCP, the default hostname "fedora" is a symptom of a problem -- the DHCP received hostname is too long and NM refuses to set it. Afterburn can write the FQDN to `/etc/hostname` and the MCO templates and the Kublet can all truncate the hostname accordingly. What we really need on FCOS on GCP is to make sure that the hostname is _set_.

**Comment by dustymabe** [comment]
> Comment [#649 (comment)](https://github.com/coreos/fedora-coreos-tracker/issues/649#issuecomment-718148008) does not address the immediate pain in GCP.
> 
> On GCP, the default hostname "fedora" is a symptom of a problem -- the DHCP received hostname is too long and NM refuses to set it. 

Having NM handle this case and truncate rather than refusing (as suggested in https://github.com/coreos/fedora-coreos-tracker/issues/649#issuecomment-718148008) would help?

> Afterburn can write the FQDN to `/etc/hostname` and the MCO templates and the Kublet can all truncate the hostname accordingly. What we really need on FCOS on GCP is to make sure that the hostname is _set_.

A combination of https://github.com/coreos/afterburn/pull/512 and https://github.com/coreos/afterburn/issues/509 would also alleviate the issue then?

**Comment by bdurrow** [comment]
In my environment the hostname returned by GCP fits in hostname.  It is still a problem there for OKD because NM dhcp is explicitly disabled by the machineconfig provided ignition configuration.  It is relatively straight forward to modify NM's DHCP behavior to modify the DHCP hostname to the desired short name.  Based on prior discussion I believe we are not willing to go that route.

**Comment by dustymabe** [comment]
> Based on prior discussion I believe we are not willing to go that route.

Any more context there or a link to where the discussion happened?

**Comment by darkmuggle** [comment]
Ah right @bdurrow
> It is still a problem there for OKD because NM dhcp is explicitly disabled by the MCO provided ignition configuration.

This is true for the MCO. For FCOS, it is not the default. And there's a _really good reason_ why this is disabled -- NM gets cantankerous when the hostname is set outside of NetworkManager and it will reset the hostname to the long hostname -- and that breaks OKD/OCP.

**Comment by jlebon** [comment]
We discussed this again in today's meeting:

```
#agreed we will ask NM whether they're willing to truncate the DHCP hostname in-tree, and if so we can work on carrying short-term script (or resort to afterburn logic for this) until that lands
```

@lucab will approach the NM developers about truncating the hostname in-tree. @cgwalters brought up that truncating might not be the right approach universally, but there's agreement that at least on GCP we do want it.

**Comment by lucab** [comment]
> @cgwalters brought up that truncating might not be the right approach universally, but there's agreement that at least on GCP we do want it.

This was related to the fact that infrastructure orchestrators may generally need to maintain routable identities for nodes based on their FQDN. That is, the local hostname should no be maimed in a way that it cannot be matched anymore at infra-level name resolution.

However if NM follows the path of https://github.com/systemd/systemd/pull/7616 I think we are not going to break that property. This is based on the fact that each label in a DNS-valid FQDN can have at most 63 chars, which is smaller than the `HOST_MAX_LEN` in Linux (64 char). Thus, "truncate overlong names to the first dot or to HOST_MAX_LEN, whatever comes earlier" will not surprisingly truncate a label mid-way even as part of long FQDN, as a single label always fits into a kernel hostname max length.

**Comment by lucab** [comment]
> @lucab will approach the NM developers about truncating the hostname in-tree

Upstream ticket at https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/572. I did reproduce the "overlong DHCP hostname" case with a customized scenario on top of `dnsmasq` and `qemu`, full NM service logs are attached there.

**Comment by LorbusChris** [comment]
Just to note: This is a release blocker for OKD 4.6

**Comment by bdurrow** [comment]
Forgive me if this is a naive question but if systemd dhcp handles this already and NM doesn't why don't we use systemd?  It looks like one of the proposals is to adopt the method that systemd already uses. It feels like NM's feature set is more closely aligned with workstations.

**Comment by cgwalters** [comment]
OK a few things here.

### Generic structural fixes

First, we should have done this as a hackmd doc with explicit listed options and their tradeoffs, rather than splitting the discussion across IRC, multiple GH PRs and issues, Slack, etc.

Second, I think we should have not tried to discuss this just inside FCOS/CoreOS team and then take it to NM as a finished design - the NM developers should have been involved in the discussion from the start once we concluded the fix should live there.

### My mistakes

Next, I apologize for missing [this comment](https://github.com/coreos/fedora-coreos-tracker/issues/649#issuecomment-722442426) which was trying to address my concerns.  

### This issue

That said while it helps I still find myself feeling that overall we're creating a potentially similar situation that led to this whole discussion - where the Fedora Workstation people changed the hostname for all Fedora systems and that unexpectedly broke OpenShift/OKD.   

Here our proposed fix for a problem that we are only hitting on GCP (right?) affects DHCP behavior across all platforms.  I feel like such a change should require a lot of consensus, probably including other clients and operating systems - i.e. this behavior should be part of an RFC or so.

There's also the sub-thread here around the operating system and "failure observability" - previously an incorrect DHCP configuration would fail in a somewhat obvious way, the hostname would stay `localhost`.  Now I think I'd agree that the "first dot" truncation probably could only help, but I'm a lot more skeptical of the "just cut it off" if there's no dot before 63-64 logic.

**Comment by lucab** [comment]
@cgwalters no worries. I also made a sad mistake in my bug report to NM, as I totally failed to grok and record @darkmuggle's important corner-case (on 64chars FQDNs). That snowballed into further communication troubles, which was pretty bad, and I'm sorry for that.

We tried to sort the details today, asking guidance to the NM folks. We summarized the two problems we have (kernel length limit and kubelet length limit) and some possible approeaches at https://hackmd.io/u2S7nEzfQMSoXKnyMEZkrA, which also contains notes from our live discussion and a tentative plan for the future.

As a way forward from here, we will continue with https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/merge_requests/677 which already solves the kernel length limit (and is enabled by default). Additionally, we will make the truncation behavior configurable via a knob, so that OKD (at least on Google cloud) can request stricter truncation.
This should allow to get the whole system back in a state where there is no need to involve dispatcher scripts, and NM can keep owning hostname handling according to the configuration suitable for higher-level systems (if any).

**Comment by dustymabe** [comment]
OK, another (probably more common) case is setting your hostname based on reverse DNS lookups. NM has code for doing the reverse DNS lookup to set the hostname but in `1.26.4` (the version in f33 right now) it priortizes reverse DNS lookup last:

https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/d256565e57af6e7f707be3f0d415eb2f225c4b78/src/nm-policy.c#L728-735

In the current case it won't fall through to reverse DNS lookup if the hostname is something other than `localhost` (or equivalent).

**Comment by dustymabe** [comment]
We discussed this at [the community meeting yesterday](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/FMTFMHSOFWRU2W2FGED3OVNGRRDVYDU5/).

We agreed to hold this week's stable release (the switch to F33) pending a discussion with the systemd team:

```
* dustymabe and jlebon to talk to the systemd team to see if we can come
  to agreement on a path forward for the fallback hostname behavior
  introduced in systemd in fedora 33
```

**Comment by dustymabe** [comment]
The maintainers have [agreed to revert the change in Fedora 33](https://bugzilla.redhat.com/show_bug.cgi?id=1892235#c19) and work to reintroduce the change at a later time once more kinks have been worked out.

**Comment by dustymabe** [comment]
New curve ball. The [revert](https://src.fedoraproject.org/rpms/systemd/c/0734f6bacd9128f4a393e526e4581844c68510cf?branch=f33) ([rpm build](https://koji.fedoraproject.org/koji/buildinfo?buildID=1656512)) didn't solve the problem completely so rather than churn [the revert was reverted](https://src.fedoraproject.org/rpms/systemd/c/8d0eb549dfd74bbca7fe495f1cba98381b725a3c?branch=f33).

This problem is extremely complex. There were two changes introduced in Fedora 33.

- the fallback hostname change in systemd [BZ#1892235](https://bugzilla.redhat.com/show_bug.cgi?id=1892235)
- the [change to systemd-resolved being enabled by default](https://fedoraproject.org/wiki/Changes/systemd-resolved)

When we revert the fallback hostname change we now break through to attempt reverse DNS because the previous things failed:

- static hostname
- from dhcp
- transient hostname (now `localhost` instead `fedora`)

So that's good, now we're trying reverse DNS again like we were in Fedora 32. 


Now's where it gets tricky and new problems arise because of the switch to `systemd-resolved` being enabled.

The `/etc/nsswitch.conf` in Fedora 33 is now 

```
hosts: files resolve [!UNAVAIL=return] myhostname dns
```

where in Fedora 32 it was:

```
hosts:      files dns myhostname
```

This means things will now go through `nss-resolve`, then `nss-myhostname`, and finally `nss-dns`. The good news is that **if** you have `systemd-resolved` enabled and reverse DNS succeeds to get an answer from an actual DNS server then the hostname will be set to the right thing. **THIS IS PROGRESS** over what we had before where reverse DNS was never tried.

However the new problems arise when either:

- there is no answer to the reverse DNS query

In this case `systemd-resolved`'s behavior is to return the fallback hostname for a reverse DNS query for the current system's IP. Now that we've changed the fallback the answer should be `localhost`... but wait.. `systemd-resolved` has a [fallback for the fallback](https://github.com/systemd/systemd/blob/9e36b885b894f1b2a53fa234a238e736582021c4/src/resolve/resolved-manager.c#L444-L453).  It basically will never return `localhost` and chooses to return `linux`.  So `systemd-resolved`'s answer to the reverse DNS lookup is `linux` and our hostname will get set to `linux`.

- the user is not using `systemd-resolved`

In this case the nsswitch.conf file falls through to the next entry after `resolve`, which is now `myhostname`, which answers `localhost`.  So our hostname will be set to `localhost` even if using `nss-dns` would have given us a different answer.





I think the most ideal thing to do in Fedora 33 is to:

1. see if we can have `systemd-resolved` stop returning a fallback to the fallback
2. reorder nsswitch.conf so that `myhostname` is lowest priority again

For after Fedora 33 we need to be [smarter in NM about the glibc resolver and possible configurations there](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/601) and also collaborate  between systemd and NetworkManager on what this should look like in the future.

**Comment by dustymabe** [comment]
I'll throw out the other options I see in addition to the one mentioned above.

The [revert was reverted](https://src.fedoraproject.org/rpms/systemd/c/8d0eb549dfd74bbca7fe495f1cba98381b725a3c?branch=f33) because:


```
    Revert the fallback hostname revert
    
    Sadly, this does not work.
    
    It seems NM queries resolved for the local IP address and gets "linux"
    and sets that as the transient hostname. Resolved has a "fallback hostname"
    (that will now again be "fedora"), but it also has a fallback fallback hostname
    that is "linux" that it used in reverse dns queries and such. NM gets
    the "linux" name and tells hostnamed to use that as the transient hostname.
    I don't think this is an improvement, since "linux" is a problematic
    as "fedora". So let's revert this for now to avoid pointless churn,
    until we figure out a real solution.
```

So theoretically completely unconfigured systems now have `linux` as their hostname. Not `localhost`, as intended by the revert of the fallback hostname change, nor `fedora` as was the case before the revert. We decided that because we got `linux` instead of `localhost` as intended, we'd not make the change for now.

The way I see it our options are to:

[1. IDEAL OPTION]

- see if we can have `systemd-resolved` stop returning a fallback to the fallback
- reorder nsswitch.conf so that `myhostname` is lowest priority again


[2. OK OPTION]

- patch the `systemd-resolvd` in fedora to use `fedora` instead of `linux` as its fallback to the fallback
- reorder nsswitch.conf so that `myhostname` is lowest priority again


[3. WORST OPTION]

- pick up `systemd-246.7-1.fc33` (has the fallback hostname change fixed)
- disable `systemd-resolved`
- change back `nsswitch.conf` to the f32 settings

This option basically reverts everything to how it was in Fedora 32

**Comment by dustymabe** [comment]
Here is another option:

 [4. SLIGHTLY LESS WORSE OPTION]
 - pick up systemd-246.7-1.fc33 (has the fallback hostname change fixed)
 - neuter `systemd-resolved` but leave the systemd-resolved systemd unit enabled
     - drop in an override that sets `DNSStubListener=no`
         - this means the real DNS servers will get used
     - change back nsswitch.conf to the f32 settings


This option is not great either, but it is attractive because of what we've already done to `testing` and `next`. In those branches we've already [ran a migration script]() to update `/etc/resolv.conf` to point to `/run/systemd/resolve/stub-resolv.conf`. Rather than have to figure out how to reverse engineer that, and the fact that we know we do want to go with systemd-resolved once this mess is figured out, we'll just continue to use systemd-resolved, but effectively have it non-functional. The only job systemd-resolved will do is create `/run/systemd/resolve/resolv.conf` and then symlink `/run/systemd/resolve/stub-resolv.conf` (which is pointed to by `/etc/resolv.conf`) to it.

**Comment by dustymabe** [comment]
ok attempt at [option 4](https://github.com/coreos/fedora-coreos-tracker/issues/649#issuecomment-743219353) has been started over in https://github.com/coreos/fedora-coreos-config/pull/780. Still pending some things that need to land first but we're almost there.

**Comment by jlebon** [comment]
The latest on this is:
- https://github.com/coreos/fedora-coreos-config/pull/868 will allow us to keep the default hostname as `localhost` without having to pin systemd
- we still need to keep neutering systemd-resolved for now
- hopefully, eventually we should be able to drop all those hacks once the NM <--> systemd bits are ironed out (https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/601)

See also https://bugzilla.redhat.com/show_bug.cgi?id=1892235#c25 for more details.

But anyway, at least as far as this specific issue is concerned, let's consider it fixed by https://github.com/coreos/fedora-coreos-config/pull/868.

**Comment by dustymabe** [comment]
followup in #834
